spec/haskell/src/SEL4/Model/StateData.lhs

```diff
@@ -332,6 +332,17 @@ The function "whileM" allows writing while loops where the loop body and the con
 > whileM :: Monad m => m Bool -> m () -> m ()
 > whileM test body = whenM test $ body >> whileM test body
 
+The function "whileLoop" emulates the definition whileLoop from NonDetMonad.thy in lib.
+
+> funOfM :: MonadState s m => m a -> s -> a
+> funOfM f s = let (rv,s') = runState f s in rv
+
+> condition :: MonadState s m => (s -> Bool) -> m a -> m a -> m a
+> condition cond left right = get >>= \s -> if (cond s) then left else right
+
+> whileLoop :: Monad m => (a -> b -> bool) -> (a -> m b) -> a -> m b
+> whileLoop cond body r = (condition (cond r) (body r) (return r)) >> whileLoop C B
```

Comment:
Does this go through Haskell translator? `Monad m` is very generic, I'm interested if this will work...

Reply:
It's ignored.

Though I've just remembered that I need to test the HaskellKernel, and that fails. Great. I hate Haskell

Reply:
Ah, yes, that one is also still open. If you're at it, see if you can make the type of `funOfM` conform to the one in Isabelle, then you don't have to exclude `awaken` from the translation.

Reply:
If you need help with that one, let me know!